File: lib/shim/vibrato.cpp
#include <stdio.h>
#include <assert.h>
#include <math.h>

#ifdef STANDALONE
#include <algorithm>
#include "AudioFile.h"
#endif

#include "vibrato.h"

float sinc(float x) {
    // TODO how on earth does this work?
    float small = 0.000000000000000000000000000000000000000000001;
    if (x < small || x > -small) {
        return 1.0;
    } else {
        return sin(x) / x;
    }
}

void Vibrato::cpp_process_audio(
    const float *input_slice,
    float *output_slice,
    size_t size,
    Playhead playhead) {
  for (size_t i=0; i < size; ++i) {
      cbuf.push(input_slice[i]);
      double tis = playhead.time_in_seconds();
      double vibrato_deviation = sin(
          tis * ((double) vibrato_frequency) * 2.0 * ((double) M_PI)) * ((double) max_sample_deviation);
      // Fractional playhead
      float fph = ((float) now_index) + vibrato_deviation;
      float window_low_f = fph - ((float) NUM_SINC_TAPS_ONE_SIDE);
      float window_high_f = fph + ((float) NUM_SINC_TAPS_ONE_SIDE);
      assert(window_low_f > 0.0);
      assert(window_high_f < (float) buffer_length);
      size_t window_low_i = (size_t) ceil(window_low_f);
      size_t window_high_i = (size_t) floor(window_high_f);
      assert(window_low_i < window_high_i);
      float convolution_sum = 0.0;
      for (size_t si = window_low_i; si < window_high_i+1; ++si) {
          float sinc_x = fph - ((float) si);
          float sinc_value = sinc(sinc_x);
          float si_sample = cbuf.get(si);
          convolution_sum += sinc_value * si_sample;
      }
      convolution_sum /= 2.0;
      if (!(convolution_sum <= 1.0 && convolution_sum >= -1.0)) {
          printf("Overflow %f\n", convolution_sum);
      }
      output_slice[i] = convolution_sum;
      playhead.inc();
  }
}

#ifdef STANDALONE
int main(int argc, char *argv[]) {
  printf("float is %lu\n", sizeof(float));
  printf("double is %lu\n", sizeof(double));
  printf("int is %lu\n", sizeof(int));

  char *infile = argv[1];
  char *outfile = argv[2];

  AudioFile<float> audioFile;
  audioFile.load(infile);

  assert(audioFile.isMono());
  assert(audioFile.getBitDepth() == 16);

  int numSamples = audioFile.getNumSamplesPerChannel();
  assert(numSamples > 0);
  int bufSize = 4;

  AudioFile<float> outAudioFile;
  outAudioFile.setAudioBufferSize(1, numSamples);
  outAudioFile.setBitDepth(audioFile.getBitDepth());
  outAudioFile.setSampleRate(audioFile.getSampleRate());
  AudioFile<float>::AudioBuffer buffer;
  buffer.resize(1);
  buffer[0].resize(numSamples);
  outAudioFile.setAudioBuffer(buffer);

  Vibrato vibrato(400, 1.0);
  Playhead playhead;

  int current = 0;
  while (current < numSamples) {
    float *inp = &audioFile.samples[0][current];
    float *outp = &outAudioFile.samples[0][current];
    size_t remaining = std::min(numSamples - current, bufSize);
    vibrato.cpp_process_audio(inp, outp, remaining, playhead);
    playhead.increment_samples(remaining);
    current += remaining;
  }

  outAudioFile.save(outfile);
}
#endif

File: lib/shared/src/filter/vibrato.rs
extern crate alloc;
extern crate libm;
#[cfg(feature = "for_host")]
extern crate std;

use alloc::boxed::Box;
use core::f32::consts::PI;

use crate::ds::circbuf::CircBuf;
use crate::knob::Knobs;
use crate::patch::Patch;
use crate::playhead::Playhead;

//#[cfg(feature = "for_host")]
//use std::println;

// The sinc convolution window is twice this.
const NUM_SINC_TAPS_ONE_SIDE: usize = 3;

// Add this many samples on either side to prevent under/overruns in production. Should
// pass rigorous testing with this set to 0, though.
const GUARD_SAMPLES: usize = 1;

fn sinc(x: f32) -> f32 {
    // TODO how on earth does this work?
    let small = 0.000000000000000000000000000000000000000000001;
    if x < small || x > -small {
        1.0
    } else {
        libm::sinf(x) / x
    }
}

pub struct Vibrato {
    // The fractional playhead can only deviate from the regular one by this much on either side.
    // Range is *exclusive*. -- TODO ??
    max_sample_deviation: usize,
    // Hz
    vibrato_frequency: f32,

    buffer_length: usize,
    now_index: usize,
    cbuf: CircBuf::<f32>,
}

impl Vibrato {
    pub fn new(max_sample_deviation: usize, vibrato_frequency: f32) -> Vibrato {
        let buffer_length: usize = 2 * (max_sample_deviation + NUM_SINC_TAPS_ONE_SIDE + GUARD_SAMPLES) + 1;
        let now_index: usize = max_sample_deviation + NUM_SINC_TAPS_ONE_SIDE + GUARD_SAMPLES;

        let vibrato = Vibrato {
            max_sample_deviation: max_sample_deviation,
            vibrato_frequency: vibrato_frequency,
            buffer_length: buffer_length,
            now_index: now_index,
            cbuf: CircBuf::<f32>::new(buffer_length, 0.0)
        };
        vibrato
    }
}

impl Patch for Vibrato {
    fn rust_process_audio(
        &mut self,
        input_slice: &[f32],
        output_slice: &mut [f32],
        _knobs: &Box<dyn Knobs>,
        mut playhead: Playhead,
    ) {
        for i in 0..input_slice.len() {
            self.cbuf.push(input_slice[i]);
            let tis = playhead.time_in_seconds();
            let vibrato_deviation = libm::sinf(
                tis * self.vibrato_frequency as f32 * 2.0 * PI as f32) * (self.max_sample_deviation as f32);
            // Fractional playhead
            let fph = (self.now_index as f32) + vibrato_deviation as f32;
            let window_low_f = fph - (NUM_SINC_TAPS_ONE_SIDE as f32);
            let window_high_f = fph + (NUM_SINC_TAPS_ONE_SIDE as f32);
            assert!(window_low_f > 0.0);
            assert!(window_high_f < self.buffer_length as f32);
            let window_low_i = libm::ceilf(window_low_f) as usize;
            let window_high_i = libm::floorf(window_high_f) as usize;
            assert!(window_low_i < window_high_i);
            let mut convolution_sum: f32 = 0.0;
            for si in window_low_i..(window_high_i+1) {
                let sinc_x = fph - (si as f32);
                let sinc_value = sinc(sinc_x);
                let si_sample = self.cbuf.get(si);
                convolution_sum += sinc_value * si_sample;
            }
            convolution_sum /= 2.0;
            /*
#[cfg(feature = "for_host")]
            if !(convolution_sum <= 1.0 && convolution_sum >= -1.0) {
                println!("Overflow {}", convolution_sum);
            }
            */
            output_slice[i] = convolution_sum;
            playhead.inc();
        }
    }
}
